### PR TITLE
JCRVLT-544 never create intermediate nodes of type nt:base or

### PR DESCRIPTION
**Pull request by kwin**
nt:hierarchyNode

Both types have too many restrictions to be useful as intermediate node
types

### CODE REVIEW

**Comment by kwin** [comment]
@DominikSuess Can you please add some ITs to verify this fixes your issue? In my AEM Quickstart SDK I don't see a package named `cq-integration-searchpromote-content`....

File: vault-core/src/main/java/org/apache/jackrabbit/vault/fs/impl/io/FolderArtifactHandler.java [hunk 1]
```diff
@@ -62,6 +70,18 @@ public void setNodeType(String nodeType) {
         this.nodeType = nodeType;
     }
 
+    private Node createIntermediateNode(Node parent, String intermediateNodeName) throws RepositoryException {
+        // preferably use default (=primary) node type for intermediate nodes
+        NodeType primaryNodeType = parent.getPrimaryNodeType();
+        final Node node;
+        if (!DISALLOWED_PRIMARY_NODE_TYPE_NAMES.contains(primaryNodeType.getName()) &&
```
Comment (DominikSuess) [inline vault-core/src/main/java/org/apache/jackrabbit/vault/fs/impl/io/FolderArtifactHandler.java:77]: @kwin  - this doesn't solve the issue at hand - it is not the primaryType of the parent that is the problem but the default nodetype - therefore the default nodetype would need to be detected and compared instead (IIUC thats part of the NodeDefinitions that can be obtained via getChildNodeDefinitions())

**Comment by DominikSuess** [comment]
@kwin  - I can provide a package isolated to the troublesome case (containing an equivalent CND & content structure) later today as I have some other high prio tasks to finish before leaving for PTO end of the week.

**Comment by kwin** [comment]
You need to, because I don't understand the issue from the description then...

**Comment by DominikSuess** [comment]
I'm a bit behind my tasks - I hope to complete by today but here the structure
Node /libs/foo with NodeType my:test with ChildeNodeDefinition definig default primary type as nt:base
Node /libs/foo/bar in package with out nodetype definition (formerly nt:folder but gets created as nt:base now)
Node /libs/foo/bar/failed  with any kind of nodetype - fails on commit as /libs/foo/bar can't take childNodes

**Comment by jsedding** [comment]
@kwin I have created a branch (see commit 7ed67c4f) with a slightly refactored `EffectiveNodeType` util. Most notably, I changed the return types of some methods to `Optional` and removed the `Predicate` arguments on the methods. Consumers of the API can filter the resulting `Optional` with custom predicates instead.

Feel free to merge or ignore as it's mostly a matter of taste.

**Comment by kwin** [comment]
@jsedding Thanks for the suggestion. I added some comments to the commit. The timing was just a bit unfortunate as I was still reworking some aspects of my branch. Now I should be done, so either you propose the changes again on top of my branch, or you wait until it is merged and then propose a PR.

**Comment by jsedding** [comment]
@kwin sorry, I won't be able to work on the proposed changes in the near future. SO please go ahead with the merge.

**Pull request merged**